### PR TITLE
fix(deployment): keyset-paginate gpu-prices query to drop deployment seq scan

### PR DESCRIPTION
**Pull request by baktun14**
## Why

`GET /v1/gpu-prices` is the #1 Postgres query by total DB load (~267K calls/24h), driving CPU spikes. Its repository query, `DeploymentRepository.findAllWithGpuResources`, prefetched every single-GPU deployment id, then re-fetched them in 200-id batches via `WHERE deployment.id IN (...)`. Per the GCP query plan, that predicate **seq-scanned the entire `deployment` table on every batch** (~309 ms × ~1,291 batches/refresh) — the dominant cost. Everything else (message/block/transaction lookups) was already sub-millisecond.

## What

- **Rewrote `findAllWithGpuResources`** to a keyset-paginated, indexed scan:
  - Pages over `deployment(createdHeight, id)` (leading columns of the existing `deployment_created_height_closed_height` index) — `ORDER BY … LIMIT` forces an index scan, eliminating the seq scan and folding away the ID-prefetch query.
  - Expresses the `gpuUnits=1` membership as a non-fanning `EXISTS` (an INNER JOIN to `deploymentGroupResource` would double-count bids for deployments with >1 GPU resource).
  - Loads bid messages via `separate: true` (a dedicated query keyed on the indexed `message.relatedDeploymentId`), so `LIMIT` counts deployments and the cursor stays exact.
  - **Output is byte-identical** — the consumer only accumulates bids and aggregates set-wise.
- **Added index** `deployment_group_deployment_id` on `deploymentGroup(deploymentId)` (an FK with no index) via **indexer migration `0008`**, so the `EXISTS` probe is index-driven. Uses `CREATE INDEX CONCURRENTLY` (non-blocking on prod) following the existing `COMMIT;--> statement-breakpoint` pattern.

### Migration safety
`0008` is `CREATE INDEX CONCURRENTLY IF NOT EXISTS` — non-blocking and idempotent. Verified it applies cleanly and creates the index in a fresh DB.

### Tests
- New `deployment.repository.integration.ts`: EXISTS double-count guard, keyset chunk-boundary (equal-`createdHeight` ties), and a no-bid-mid-chunk case proving paging doesn't terminate early.
- Existing functional exact-JSON test for `GET /v1/gpu-prices` stays green (output equivalence).

### Validated on prod (`prod-mainnet`, read-only EXPLAIN; index created `CONCURRENTLY` then dropped)

`EXPLAIN (ANALYZE, BUFFERS)` per keyset chunk on real `akash-mainnet` data (`createdHeight ≥ ~31d`, chunk 1000):

| Query | Plan | Exec |
|---|---|---|
| OLD `id IN (200)` batch | Seq Scan on `deployment` | **555 ms** × ~1,291 batches/refresh |
| NEW keyset chunk — **no index** | `EXISTS` → hash semi join over full `deploymentGroup` (1.44M) + `deploymentGroupResource` (1.26M) scans + top-N sort | **7,638 ms** |
| NEW keyset chunk — **with `deployment_group_deployment_id`** | `EXISTS` → Nested Loop Semi Join via the index; index scan on `createdHeight`; `LIMIT` short-circuits (~7K rows, not 306K) | **346 ms** |

⇒ The `0008` index is **load-bearing, not optional**: without it the rewrite is slower than today; with it, ~22× faster per chunk and the `LIMIT` short-circuits. **Deploy ordering:** the index is created by the indexer's 0008 migration, so the indexer should deploy before/with the api or there's a transient window of slow background refreshes (tolerable — `/v1/gpu-prices` is a memoized 15-min background refresh, p75 31 ms cache hits in Sentry).

**Output equivalence on prod data:** the new `EXISTS` selects the **identical 262,362 GPU deployments** as the old join (`DISTINCT`), with the same bids. 159 deployments (0.06%) have >1 `gpuUnits=1` resource; the old path could double-count their bids across 200-batches, the new yields each once — but that only affected the debug-only `bidCount`, never the public `price`/`priceUakt`/`availability` (best-bid-per-provider aggregation is duplicate-insensitive). **Public output unchanged.**

## Summary by CodeRabbit

* **New Features**
  * Configurable keyset pagination for GPU deployment queries with chunk-size validation.

* **Performance Improvements**
  * Keyset-based scan and a new DB index to speed GPU-resource membership checks and scalable enumeration.

* **Bug Fixes**
  * Ensure deployments with only non-qualifying GPU units or with creation before minHeight are excluded; avoid duplicate results.

* **Tests**
  * Integration tests covering filtering rules, bid-association, deterministic pagination, and boundary/edge cases.

### CODE REVIEW

File: apps/api/src/deployment/repositories/deployment/deployment.repository.ts [hunk 1]
```diff
@@ -104,61 +104,74 @@ export class DeploymentRepository {
...
-    const ids = recordsWithIds.map(r => r.id);
-    const BATCH_SIZE = 200;
-
-    // Iterate over IDs in batches and yield each deployment
-    for (let i = 0; i < ids.length; i += BATCH_SIZE) {
-      const batchIds = ids.slice(i, i + BATCH_SIZE);
-
+  async *findAllWithGpuResources(options: { minHeight: number; chunkSize?: number }) {
+    const BID_TYPES = ["/akash.market.v1beta4.MsgCreateBid", "/akash.market.v1beta5.MsgCreateBid"];
+    const chunkSize = options.chunkSize ?? 1000;
```
[the remark this answers is not in the document]
Comment (baktun14) [reply apps/api/src/deployment/repositories/deployment/deployment.repository.ts:109]: Fixed in 1378d6a54: `chunkSize` is now validated as a positive integer (throws otherwise), with a test. The single check covers both the missed empty-batch break and the `batch[-1]` dereference.

**Review by stalniy — APPROVED** [review]
